### PR TITLE
Gpu datatest

### PR DESCRIPTION
**Pull request by OngChia**
Enable single-node GPU run for the Jablonowski-WIlliamson test on R2B4 grid and all datatests for the Timeloop.
1. Change np to xp in `serialbox_utils.py`.
2. Change asnumpy() to ndarray in `test_timeloop.py` and `jablownski_williamson.py` and `driver/utils.py`
3. Split `predictor_stencils_7_8_9` to two stencils to avoid illegal memory access when running on gpu.
4. Add a new field `exner_dyn_incr_lastsubstep` in class `DiagnosticStateNonHydro` as the output argument for the stencil `update_dynamical_exner_time_increment` because this stencil originally updates the input argument `exner_dyn_incr`.

### CODE REVIEW

**Comment by OngChia** [comment]
cscs-ci run default

File: model/common/src/icon4py/model/common/decomposition/definitions.py [hunk 1]
```diff
@@ -69,6 +70,7 @@ class EntryType(IntEnum):
     @builder.builder
     def with_dimension(self, dim: Dimension, global_index: np.ndarray, owner_mask: np.ndarray):
         masked_global_index = ma.array(global_index, mask=owner_mask)
+        masked_global_index = xp.asarray(masked_global_index)
         self._global_index[dim] = masked_global_index
```
Comment (halungge) [inline model/common/src/icon4py/model/common/decomposition/definitions.py:74]: I guess that won't work. I don't now what happens with the mask if you do that and we need it that mask later. But indeed cupy does not implement the numpy's masked_array interface. We would have to do a simple hand made implementation which should be enough for what we need.   I'll open a fix for that.

**Review by halungge — CHANGES_REQUESTED** [review]
Some renaming, for the numpy.ma parts you might pick changes from this [PR](https://github.com/C2SM/icon4py/pull/554)

File: model/driver/src/icon4py/model/driver/test_cases/jablonowski_williamson.py [hunk 2]
```diff
@@ -33,6 +33,7 @@
 log = logging.getLogger(__name__)
 
 
+# TODO (Chia Rui): Convert all numpy computations to cupy
```
[the remark this answers is not in the document]
Comment (OngChia) [reply model/driver/src/icon4py/model/driver/test_cases/jablonowski_williamson.py:36]: Ah. Thanks. I forgot to delete it.

**Comment by OngChia** [comment]
cscs-ci run default

**Comment by OngChia** [comment]
cscs-ci run default

**Comment by OngChia** [comment]
launch jenkins spack

**Comment by OngChia** [comment]
cscs-ci run default

**Comment by OngChia** [comment]
launch jenkins spack

**Comment by OngChia** [comment]
cscs-ci run default

**Comment by OngChia** [comment]
launch jenkins spack

**Comment by OngChia** [comment]
cscs-ci run default

**Comment by OngChia** [comment]
launch jenkins spack

**Pull request closed without merging**